nomos-services/cryptarchia-consensus/src/states.rs

```diff
@@ -0,0 +1,132 @@
+// std
+use std::marker::PhantomData;
+// Crates
+use cl::NoteWitness;
+use overwatch_rs::services::state::ServiceState;
+use serde::{Deserialize, Serialize};
+// Internal
+use crate::leadership::Leader;
+use crate::{Cryptarchia, CryptarchiaSettings, Error};
+use nomos_core::header::HeaderId;
+use nomos_ledger::LedgerState;
+
+pub struct GenesisRecoveryStrategy {
+    pub tip: HeaderId,
+}
+
+pub struct SecurityRecoveryStrategy {
+    pub tip: HeaderId,
+    pub security_block_id: HeaderId,
+    pub security_ledger_state: LedgerState,
+    pub security_leader_notes: Vec<NoteWitness>,
+}
+
+pub enum CryptarchiaInitialisationStrategy {
+    Genesis,
+    RecoveryFromGenesis(GenesisRecoveryStrategy),
+    RecoveryFromSecurity(Box<SecurityRecoveryStrategy>),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct CryptarchiaConsensusState<TxS, BxS, NetworkAdapterSettings, BlendAdapterSettings> {
+    tip: Option<HeaderId>,
+    security_block: Option<HeaderId>,
+    security_ledger_state: Option<LedgerState>,
+    security_leader_notes: Option<Vec<NoteWitness>>,
+    _txs: PhantomData<TxS>,
+    _bxs: PhantomData<BxS>,
+    _network_adapter_settings: PhantomData<NetworkAdapterSettings>,
+    _blend_adapter_settings: PhantomData<BlendAdapterSettings>,
+}
+
+impl<TxS, BxS, NetworkAdapterSettings, BlendAdapterSettings>
+    CryptarchiaConsensusState<TxS, BxS, NetworkAdapterSettings, BlendAdapterSettings>
+{
+    pub fn new(
+        tip: Option<HeaderId>,
+        security_block: Option<HeaderId>,
+        security_ledger_state: Option<LedgerState>,
+        security_leader_notes: Option<Vec<NoteWitness>>,
+    ) -> Self {
+        Self {
+            tip,
+            security_block,
+            security_ledger_state,
+            security_leader_notes,
+            _txs: Default::default(),
+            _bxs: Default::default(),
+            _network_adapter_settings: Default::default(),
+            _blend_adapter_settings: Default::default(),
+        }
+    }
+
+    pub(crate) fn from_cryptarchia(cryptarchia: &Cryptarchia, leader: &Leader) -> Self {
+        let security_block_header = cryptarchia.consensus.get_security_block_header_id();
+        let security_ledger_state = security_block_header
+            .and_then(|header| cryptarchia.ledger.state(&header))
+            .cloned();
+        let security_leader_notes = security_block_header
+            .and_then(|header_id| leader.notes(&header_id))
+            .cloned();
+
+        Self::new(
+            Some(cryptarchia.tip()),
+            security_block_header,
+            security_ledger_state,
+            security_leader_notes,
+        )
+    }
+
+    fn can_recover(&self) -> bool {
```

Comment:
Could be a const function
```suggestion
    fn const can_recover(&self) -> bool {
```

Reply:
I wonder, how does the compiler resolve this?